### PR TITLE
2588 Request tokens from terriajs-server and plumb them into ArcGisMapServerImageryProvider.

### PR DESCRIPTION
**Pull request by a-stacey**
There is one outstanding piece of lint associated with this code. `function requestMetadata(token) {` is inline. I have left this as is for now as I wonder whether there is a better solution to this problem then to use an intermediate function. The problem that this inline function solves is to prepend the token request in the initalisation process, and so I wonder whether there is a better solution to this problem then using the if -> inline function call.

The inline function presents particular nastiness in that to comply with the lint checks this code needs to be made top level function, however this in its currently implementation is not particularly clean in that doing in as a straight copy-paste means that the checks are separated from the implementation. To clean this up the checks can be moved into the separate function, but then the token request may happen when we know the other checks will fail. So this is not particularity nice overall and so I'm hoping that there is a cleaner higher level solution so that all of this complexity can be avoided.

### CODE REVIEW

File: lib/Models/ArcGisMapServerCatalogItem.js [hunk 1]
```diff
@@ -84,7 +86,13 @@ var ArcGisMapServerCatalogItem = function(terria) {
      */
     this.allowFeaturePicking = true;
 
-    knockout.track(this, ['layers', 'maximumScale', '_legendUrl', '_generatedLegendUrl', 'maximumScaleBeforeMessage', 'showTilesAfterMessage', 'allowFeaturePicking']);
+    /**
+     * Gets or sets the URL to use for requesting tokens.
+     * @type {String}
+     */
+    this.tokenSource = undefined;
```
Comment (kring) [inline lib/Models/ArcGisMapServerCatalogItem.js]: I think `tokenUrl` would be a better name than `tokenSource`.

File: lib/Models/ArcGisMapServerCatalogItem.js [hunk 2]
```diff
@@ -195,6 +203,43 @@ function getJson(item, uri) {
...
+
+        item._lastToken = undefined;
+        throw new RuntimeError('The token server responded with an invalid token.');
+    }).otherwise(() => {
+        item._lastToken = undefined;
+        throw new RuntimeError('Unable to request a token from the token server.');
+    });
+}
+
+function appendQuery(query, addition) {
```
Comment (kring) [inline lib/Models/ArcGisMapServerCatalogItem.js]: `urijs` has a handy method to do this: http://medialize.github.io/URI.js/docs.html#search-set

File: lib/Models/ArcGisMapServerCatalogItem.js [hunk 3]
```diff
@@ -195,6 +203,30 @@ function getJson(item, uri) {
...
+        if (defined(token.token)) {
+            item._lastToken = token.token;
+            return token.token;
+        }
+
+        item._lastToken = undefined;
+        throw new RuntimeError('The token server responded with an invalid token.');
+    }).otherwise(() => {
+        item._lastToken = undefined;
+        throw new RuntimeError('Unable to request a token from the token server.');
```
Comment (kring) [inline lib/Models/ArcGisMapServerCatalogItem.js]: Probably better to throw `TerriaError` here (and above).

File: lib/Models/ArcGisMapServerCatalogItem.js (hunk 2 above)
Comment (a-stacey) [reply lib/Models/ArcGisMapServerCatalogItem.js]: Thanks for this.

f8b72f88960ec67655cc1b4805409459b3c0b78d

File: lib/Models/ArcGisMapServerCatalogItem.js (hunk 1 above)
Comment (a-stacey) [reply lib/Models/ArcGisMapServerCatalogItem.js]: Thanks much better.

8c83029b95bfccb2748606190c2018035b1bf6ed

**Comment by kring** [comment]
Just that one last comment!

File: lib/Models/ArcGisMapServerCatalogItem.js (hunk 3 above)
Comment (a-stacey) [reply lib/Models/ArcGisMapServerCatalogItem.js]: Thanks.

5bcbe3e2fe6c8272d7239ccd0169b41a30f4e581

**Pull request merged**